cmd/sftp-server.go

```diff
@@ -227,6 +243,44 @@ func startSFTPServer(args []string) {
 			MACs:         allowMACs,
 		},
 		PublicKeyAuthAlgorithms: allowPubKeys,
+		PublicKeyCallback: func(c ssh.ConnMetadata, key ssh.PublicKey) (*ssh.Permissions, error) {
+			if caPublicKey == nil {
+				return nil, errAuthentication
+			}
+			_, ok := globalIAMSys.GetUser(context.Background(), c.User())
+			if !ok {
+				return nil, errNoSuchUser
+			}
+
+			// Verify that client provided certificate, not only public key.
+			cert, ok := key.(*ssh.Certificate)
+			if !ok {
+				return nil, errAuthentication
+			}
+
+			// Verify that identity in certificate matches to username in authentication request.
+			if cert.KeyId != c.User() {
+				return nil, errAuthentication
+			}
+
+			// Extract data structure without signature from authentication message
+			// and remove last 4 bytes which represent the length of the non-existent
+			// signature field which appears there as result of Marshal()
+			signed := *cert
+			signed.Signature = nil
+			signedData := signed.Marshal()
+			signedData2 := signedData[:len(signedData)-4]
```

Comment:
why is there no cleaner way to do this? is this the expected way to verify the incoming pubkey signature ?

Reply:
Actually yes, there is [Authenticate](https://cs.opensource.google/go/x/crypto/+/refs/tags/v0.22.0:ssh/certs.go;l=358-381) function and documentation even says that it can be used with `PublicKeyCallback`

Now code most probably is more clear.